### PR TITLE
Update Graph APIs based on API review feedback

### PR DESCRIPTION
**Pull request by saeta**
Based on the round of in person discussion, we came up with a few ideas on how to improve the `PenguinGraph` API.

### CODE REVIEW

File: Sources/PenguinGraphs/DepthFirstSearch.swift [hunk 1]
```diff
@@ -27,99 +27,112 @@ public enum VertexColor {
...
+    try depthFirstSearch(
+      startingAt: startVertex, vertexVisitationState: &vertexVisitationState, visitor: &visitor)
+  }
+
+  /// Runs depth first search on starting at `startVertex` using `vertexVisitationState` to keep track of
   /// visited vertices; `visitor` is called regularly to allow arbitrary state to be computed during
   /// search.
   ///
-  /// - Note: `graph` is taken `inout` because the `colorMap` or `visitor` may store data within the
+  /// - Note: this is a mutating function because the `vertexVisitationState` or `visitor` may store data within the
```
[the remark this answers is not in the document]
Comment (saeta) [reply Sources/PenguinGraphs/DepthFirstSearch.swift:53]: +1; fixed!

File: Sources/PenguinGraphs/BreadthFirstSearch.swift [hunk 2]
```diff
@@ -14,83 +14,78 @@
...
+    VertexVisitationState.Graph == Self,
+    VertexVisitationState.Value == VertexColor,
+    StartVertices.Element == VertexId
   {
     precondition(!startVertices.isEmpty, "startVertices was empty.")
     for startVertex in startVertices {
-      colorMap.set(vertex: startVertex, in: &graph, to: .gray)
-      try visitor.start(vertex: startVertex, &graph)
-      try visitor.discover(vertex: startVertex, &graph)
+      vertexVisitationState.set(vertex: startVertex, in: &self, to: .gray)
```
[the remark this answers is not in the document]
Comment (saeta) [reply Sources/PenguinGraphs/BreadthFirstSearch.swift:63]: Agreed; unfortunately, we can't use `inout` on subscript parameters. :-(

File: Sources/PenguinGraphs/AdjacencyList.swift [hunk 3]
```diff
@@ -302,7 +302,7 @@ extension AdjacencyList: MutablePropertyGraph {
   /// Adds a new vertex with associated `information`, returning its identifier.
   ///
   /// - Complexity: O(1) (amortized)
-  public mutating func addVertex(with information: Vertex) -> VertexId {
+  public mutating func addVertex(_ information: Vertex) -> VertexId {
```
[the remark this answers is not in the document]
Comment (dabrahams) [reply Sources/PenguinGraphs/AdjacencyList.swift]: Friday: sure.  I think `properties` or `storedProperties` would be a better name in this context than `vertexProperty`.  We could also think about referring to these as “vertex contents” or “vertex values,” as long as we pick a consistent terminology.  I think the whole “properties” thing came from a time before bundled properties in BGL and the relationship to property maps.

File: Sources/PenguinGraphs/AdjacencyList.swift [hunk 4]
```diff
@@ -313,7 +313,7 @@ extension AdjacencyList: MutablePropertyGraph {
   ///
   /// - Complexity: O(1) (amortized)
   public mutating func addEdge(
-    from source: VertexId, to destination: VertexId, with information: Edge
+    from source: VertexId, to destination: VertexId, _ information: Edge
```
[the remark this answers is not in the document]
Comment (dabrahams) [reply Sources/PenguinGraphs/AdjacencyList.swift]: But `addVertex` should be consistent with this, of course!

File: Sources/PenguinGraphs/BreadthFirstSearch.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (dabrahams) [reply Sources/PenguinGraphs/BreadthFirstSearch.swift:63]: That simply silly.

File: Sources/PenguinGraphs/DepthFirstSearch.swift [hunk 5]
```diff
@@ -27,99 +27,112 @@ public enum VertexColor {
...
+  /// Runs depth first search on starting at `startVertex`; `visitor` is called regularly to allow
+  /// arbitrary state to be computed during search.
+  ///
+  /// - Note: this is a mutating function because the `visitor` may store data within the graph
+  ///   itself.
+  public mutating func depthFirstSearch<
+    Visitor: DFSVisitor
+  >(
+    startingAt startVertex: VertexId,
+    visitor: inout Visitor
```
[the remark this answers is not in the document]
Comment (dabrahams) [reply Sources/PenguinGraphs/DepthFirstSearch.swift:41]: Yeah, just groping for a more role-denoting name.  But `visitor` may actually be the best description of the role here.

**Pull request merged**